### DISCUSSION TITLE
Run parsing stats with each grammar PR?

### DISCUSSION THREAD

**Discussion by mjambon** [tree-sitter/tree-sitter#1001] [Ideas] [answered]
To track the quality of a parser used in semgrep, we've been running stats on public git repos for each target language we're supporting or consider supporting. A sample result is ["99.94% parsing success for ruby"](https://dashboard.semgrep.dev/metric/semgrep.core.ruby.parse.pct). I suggest we do the same for each tree-sitter-* grammar, automatically on each pull request.

Why?
--

Tests are sometimes incomplete, and they don't reflect the quantitative impact of changes on real programs. If a new syntax error gets through unnoticed, the consequences can be minor or major. It depends on how often the construct is used and how many lines the ERROR node affects. Knowing the consequences of a change on real code would eliminate guesswork and anxiety when reviewing and merging pull requests for tree-sitter grammars, speeding up contributions.

Method
--

* Each tree-sitter grammar repo needs to define a list of public git(hub) URLs and file extensions for the language (we have a script to help with this).
* A script takes care of cloning the repos and reporting (at least to stdout) how many lines were parsed successfully, as a fraction of the total number of lines.
* The script can be run locally by the grammar contributor.
* The script would be run as part of a CI job on pull requests. This job would always succeed and print the parsing rate such that the reviewers of the pull request can see it.
* A daily CI job would record the parsing rate (where? writing to a file in the same github repo, possibly in a special branch is possible)

Proposal
--

I can take care of this as part of the work we do for semgrep if this facilitates contributions. I would create simple instructions for setting this up and would set it up for the existing grammars.

**Comment by razzeee** [comment]
I have most of this implemented here https://github.com/elm-tooling/tree-sitter-elm/blob/main/.github/workflows/test-full.yml

but on every code change and not on a schedule

**Comment by maxbrunsfeld** [comment]
> A sample result is "99.94% parsing success for ruby". 

I love those graphs that show the success rate over time; that's very cool.

> I suggest we do the same for each tree-sitter-* grammar, automatically on each pull request.

We're already doing something similar to this on every CI build for most of the Tree-sitter grammars. Here's where we do it for the [TypeScript](https://github.com/tree-sitter/tree-sitter-python/blob/20729b3e2fe70d13c1dd9f400382871a5fe95255/script/parse-examples#L19-L21), [Python](https://github.com/tree-sitter/tree-sitter-python/blob/20729b3e2fe70d13c1dd9f400382871a5fe95255/script/parse-examples#L19-L21), and [Go](https://github.com/tree-sitter/tree-sitter-go/blob/2a83dfdd759a632651f852aa4dc0af2525fae5cd/script/parse-examples#L18-L19) parsers. The build fails if any *new* file fails to parse. If the success rate is <100%, we currently handle that by maintaining a list of files called [`known-failures`](https://github.com/tree-sitter/tree-sitter-typescript/blob/master/script/known-failures.txt), checked into the repositories, and the idea is that after adding an example repo, we never add to `known-failures`, but only remove files from it over time.

I think that for most grammars, we could increase the *number* of example repositories that we parse; right now, in a lot of places we're just doing 1 - 5 repos.

> A daily CI job would record the parsing rate (where? writing to a file in the same github repo, possibly in a special branch is possible)

Yeah, this sounds like a cool addition. I'm just  not sure how we should record and display the data.

**Reply by mjambon** [reply]
If you say the current PR checks are satisfying, maybe I won't do that. I think it would be nice to have something more reusable, i.e. download the script(s) from a dedicated repo rather having a copy in each grammar repo.

For good stats, scanning over 1-10 million lines [is preferable](https://github.com/returntocorp/ocaml-tree-sitter/blob/master/doc/parsing-stats-precision.md) as the error rate gets close to 0.01%. The problem is it can take more than 10 minutes, which is not good for getting quick feedback on a PR.

Anyway, I'm considering starting with running daily stats for all languages from a central repo. This makes it easy to have everything set up quickly without bothering every grammar maintainer.

**Reply by maxbrunsfeld** [reply]
Yeah, I agree about having something more reusable. I think that could be either a shared script, as you suggest, or even some dedicated command built into the `tree-sitter` CLI.

> For good stats, scanning over 1-10 million lines is preferable as the error rate gets close to 0.01%. The problem is it can take more than 10 minutes, which is not good for getting quick feedback on a PR.

Yeah. In our current CI builds, I don't think we have a large enough sample size. 10 minutes might be a bit extreme for running on every PR, but I think it would be totally fine if it took 1 - 3 minutes. Right now, it's typically more like a few seconds.

> Anyway, I'm considering starting with running daily stats for all languages from a central repo. This makes it easy to have everything set up quickly without bothering every grammar maintainer.
 
 Love it; this would be a great addition.

**Comment by ObserverOfTime** [comment, marked as answer]
https://github.com/tree-sitter/parse-action
